zebra/zebra_srv6.c

```diff
@@ -0,0 +1,333 @@
+/*
+ * Zebra SRv6 definitions
+ * Copyright (C) 2020  Hiroki Shirokura, LINE Corporation
+ * Copyright (C) 2020  Masakazu Asama
+ *
+ * This program is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License as published by the Free
+ * Software Foundation; either version 2 of the License, or (at your option)
+ * any later version.
+ *
+ * This program is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
+ * more details.
+ *
+ * You should have received a copy of the GNU General Public License along
+ * with this program; see the file COPYING; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA
+ */
+
+#include <zebra.h>
+
+#include "network.h"
+#include "prefix.h"
+#include "stream.h"
+#include "srv6.h"
+#include "zebra/zapi_msg.h"
+#include "zebra/zserv.h"
+#include "zebra/zebra_router.h"
+#include "zebra/zebra_srv6.h"
+#include "zebra/zebra_memory.h"
+#include "zebra/zebra_errors.h"
+#include <stdio.h>
+#include <string.h>
+#include <stdlib.h>
+#include <arpa/inet.h>
+#include <netinet/in.h>
+
+#include <stdio.h>
+#include <string.h>
+#include <stdlib.h>
+#include <arpa/inet.h>
+#include <netinet/in.h>
+
+
+static void set_seg6local_end(struct prefix_ipv6 *prefix)
+{
+	struct route_entry *re;
+	struct nexthop_group *ng;
+	struct vrf *vrf;
+	struct interface *ifp;
+	struct nexthop *nexthop;
+	struct seg6local_context ctx;
+	int ret;
+	ifindex_t ifindex;
+
+	re = XCALLOC(MTYPE_RE, sizeof(struct route_entry));
+	re->type = ZEBRA_ROUTE_STATIC;
+	re->instance = 0;
+	re->uptime = monotime(NULL);
+	re->vrf_id = VRF_DEFAULT;
+	re->table = RT_TABLE_MAIN;
+	SET_FLAG(re->flags, ZEBRA_FLAG_SEG6LOCAL_ROUTE);
+
+	ifindex = 1;
+	vrf = vrf_lookup_by_id(VRF_DEFAULT);
+	FOR_ALL_INTERFACES (vrf, ifp) {
+		if (ifp->ifindex == 1)
+			continue;
+		if (!if_is_operative(ifp))
+			continue;
+		ifindex = ifp->ifindex;
+		break;
+	}
+
+	if (ifindex == 1)
+		zlog_err("%s: no available interface on default-vrf",
+			 __func__);
+
+	ng = nexthop_group_new();
+	nexthop = nexthop_from_ifindex(ifindex, VRF_DEFAULT);
+	if (!nexthop) {
+		flog_warn(
+			EC_ZEBRA_NEXTHOP_CREATION_FAILED,
+			"%s: Nexthops Specified: but we failed to properly create one",
+			__PRETTY_FUNCTION__);
+		nexthop_group_delete(&ng);
+		XFREE(MTYPE_RE, re);
+		return;
+	}
+
+	memset(&ctx, 0, sizeof(ctx));
```

Comment:
Same thing as above ... just making certain we should be using memset here.

Reply:
I see, I'll fix without memset. as following. 

```diff
diff --git a/zebra/zebra_srv6.c b/zebra/zebra_srv6.c
index 82ff498af..8ba37bf65 100644
--- a/zebra/zebra_srv6.c
+++ b/zebra/zebra_srv6.c
@@ -50,7 +50,7 @@ static void set_seg6local_end(struct prefix_ipv6 *prefix)
        struct vrf *vrf;
        struct interface *ifp;
        struct nexthop *nexthop;
-       struct seg6local_context ctx;
+       struct seg6local_context ctx = {0};
        int ret;
        ifindex_t ifindex;

@@ -89,7 +89,6 @@ static void set_seg6local_end(struct prefix_ipv6 *prefix)
                return;
        }

-       memset(&ctx, 0, sizeof(ctx));
        nexthop_add_seg6local(nexthop, ZEBRA_SEG6_LOCAL_ACTION_END, &ctx);
        nexthop_group_add_sorted(ng, nexthop);
```

Reply:
I fixed it.

Reply:
Sorry, My above solution is not good for various compilers.
i.e. https://ci1.netdef.org/browse/FRR-FRRPULLREQ-CI011BUILD-13620/log

So, we should not use this initialization. `struct A a = {0}`.

May be, better solution is follow.
```diff
diff --git a/lib/nexthop.c b/lib/nexthop.c
index ce47730ac..cd98c429a 100644
--- a/lib/nexthop.c
+++ b/lib/nexthop.c
@@ -493,7 +493,8 @@ void nexthop_add_seg6local(struct nexthop *nexthop, uint32_t action,
                return;

        nh_ctx = XCALLOC(MTYPE_NH_SEG6LOCAL, sizeof(struct seg6local_context));
-       *nh_ctx = *ctx;
+       if (ctx)
+               *nh_ctx = *ctx;
        nexthop->nh_seg6local_action = action;
        nexthop->nh_seg6local_ctx = nh_ctx;
 }
diff --git a/zebra/zebra_srv6.c b/zebra/zebra_srv6.c
index 8ba37bf65..d886566cd 100644
--- a/zebra/zebra_srv6.c
+++ b/zebra/zebra_srv6.c
@@ -50,7 +50,6 @@ static void set_seg6local_end(struct prefix_ipv6 *prefix)
        struct vrf *vrf;
        struct interface *ifp;
        struct nexthop *nexthop;
-       struct seg6local_context ctx = {0};
        int ret;
        ifindex_t ifindex;

@@ -89,7 +88,7 @@ static void set_seg6local_end(struct prefix_ipv6 *prefix)
                return;
        }

-       nexthop_add_seg6local(nexthop, ZEBRA_SEG6_LOCAL_ACTION_END, &ctx);
+       nexthop_add_seg6local(nexthop, ZEBRA_SEG6_LOCAL_ACTION_END, NULL);
        nexthop_group_add_sorted(ng, nexthop);

        ret = rib_add_multipath(AFI_IP6, SAFI_UNICAST,
```

Reply:
I fixed it with following way
https://github.com/FRRouting/frr/pull/5865/files#diff-eefc2a635b2d5e71df8b796ea5e28b99R91
https://github.com/FRRouting/frr/pull/5865/files#diff-69588866f54e3caa134448eeb6b3b118R496-R497